docs/ppvit-config.md
## PPViT: How to use config?
> sample code: [here](https://github.com/xperzy/PPViT/blob/develop/image_classification/ViT/config.py)

This document presents the basics of `config` that used in **PPViT** project. 

The core module used in PPViT `config` is [yacs](https://github.com/rbgirshick/yacs) (0.1.8+). Similar as other projects, PPViT `config` supports loading from [yaml](https://yaml.org/) file, and configarable using python [ArgumentParser](https://docs.python.org/3/library/argparse.html).

> Full usage of `yacs` can be found in https://github.com/rbgirshick/yacs

### 1. Installation
#### 1.1 Install by `pip`
To intall `yacs` version `0.1.8`:
```shell
$ pip install yacs==0.1.8
```
#### 1.2 Install from source
You can also install `yacs` from gitub:
```shell
$ git clone https://github.com/rbgirshick/yacs.git
$ cd yacs
$ python setup.py install
```

### 2. Basic Concepts and Usage
#### 1. CfgNode
`CfgNode` represents an internal node in the configuration tree. It is a `dict`-like container, which allows attribute-based access to its keys.
```python
from yacs.config import CfgNode as CN

_C = CN()

_C.NUM_GPUS = 4
_C.NUM_WORKERS = 8
_C.BATCH_SIZE = 128

def get_config():
    return _C.clone()
```
#### 2. Read `.yaml` file using `merge_from_file()`
`yacs` allows reading YAML file to override `CfgNode`. You can create a `.yaml` file for each experiment, which will only change the options in that expriment.

Some basic format in YAML file:
```YAML
key:    # YAML uses 'key: value' paris, separated using ':'
    child_key: value    # indent can be used to show different levels
    child_KEY2: value3  # YAML is case sensitive
    c_arr: [val1, val2, val3]   # array can be used in value
    c_bool: True    # True/true/TRUE are all OK
    c_float: 3.1415 # float is allowed
    c_string: no quote is allowed # "", '', no quote are all OK
```

`merge_from_file()` can be used to override current `CfgNode`:
```python
cfg = get_config()
cfg.merge_from_file('experiment_1.yaml')
print(cfg)
```

#### 3. Override by `ArgumentParser`
You can write your own method to update config using python `ArgumentParser`, e.g.:
```python
def update_config(config, args)
    if args.cfg:    # update from .yaml file
        upate_config_from_file(config, args.cfg)
    if args.batch_size: # update BATCH_SIZE
        config.BATCH_SIZE = args.batch_size
    if args.eval:
        config.EVAL = True
    return config
```




### 4. Practical Guide of using config for PPViT:
#### STEP 1: Create config.py
Create a python file config.py, which is the place to define **all the configurable options**. It should be well documented and provide suitable default values for all options. 
Typically, `config.py` should have:
- `DATA`: defines the dataset path, input image size, and batch_size, etc.
- `MODEL`:
    - General options for your model, such as model name, num_classes, etc.
    - `TRANS`: transformer related options, such as mlp dimention, hidden dimension, number of heads, etc.
- `TRAIN`: training related options, such as num of epochs, lr, weight decay, etc.

In `config.py`, you should implement `update_config(config, args)`, which reads current `config` and `args` from `ArgumentParser` to update the config using commandline options.

#### STEP 2: 
In your `main.py`, create `ArgumentParser` which includes all the options in `update_config(config, args)` method from `config.py`, e.g.:
```python
  parser = argparse.ArgumentParser('ViT')
  parser.add_argument('-cfg', type=str, default=None)
  parser.add_argument('-dataset', type=str, default=None)
  parser.add_argument('-batch_size', type=int, default=None)
  parser.add_argument('-image_size', type=int, default=None)
  parser.add_argument('-data_path', type=str, default=None)
  parser.add_argument('-ngpus', type=int, default=None)
  parser.add_argument('-pretrained', type=str, default=None)
  parser.add_argument('-eval', action='store_true')
  args = parser.parse_args()

  # get default config
  config = get_config()
  # update config by arguments
  config = update_config(config, args)
```

Then you can use the attribute-based access to get the config option values.

#### STEP 3:
You should create a single `.yaml` file for each experiment, e.g.,
```yaml
DATA:
    IMAGE_SIZE: 224
    CROP_PCT: 0.875
MODEL:
    TYPE: ViT
    NAME: vit_large_patch16_224
    TRANS:
    PATCH_SIZE: 16
    HIDDEN_SIZE: 1024
    MLP_DIM: 4096 # same as mlp_ratio = 4.0
    NUM_LAYERS: 24
    NUM_HEADS: 16
    QKV_BIAS: True
```

If you set the command line argument `-cfg` to the `.yaml` file path, the config will be override with the file options. 
> **Note:** the `.yaml` overrides the config before the `args`, so the options in `args` are the current options.

semantic_segmentation/config.py
import os
from yacs.config import CfgNode as CN
import yaml

_C = CN()
_C.BASE = ['']

# data settings
_C.DATA = CN()
_C.DATA.BATCH_SIZE = 4  #train batch_size for single GPU
_C.DATA.BATCH_SIZE_VAL = 1 # val batch_size for single GPU
_C.DATA.DATASET = 'PascalContext' # dataset name
_C.DATA.DATA_PATH = '/home/ssd3/wutianyi/datasets/pascal_context'
_C.DATA.CROP_SIZE = (480,480) # input_size (training)
_C.DATA.NUM_CLASSES = 60  # 19 for cityscapes, 60 for Pascal-Context
_C.DATA.NUM_WORKERS = 0 # number of data loading threads (curren paddle must set to 0)

# model settings
_C.MODEL = CN()
_C.MODEL.NAME = 'SETR_MLA'
_C.MODEL.ENCODER = CN()
_C.MODEL.ENCODER.TYPE = 'ViT_MLA'
_C.MODEL.ENCODER.OUT_INDICES = [5,11,17,23]

_C.MODEL.DECODER_TYPE = 'ViT_MLAHead'
_C.MODEL.RESUME = None
_C.MODEL.PRETRAINED = None
_C.MODEL.NUM_CLASSES = 1000
_C.MODEL.DROPOUT = 0.0 # 0.0
_C.MODEL.ATTENTION_DROPOUT = 0.0

# transformer settings
_C.MODEL.TRANS = CN()
_C.MODEL.TRANS.HYBRID = False       #TODO: implement
_C.MODEL.TRANS.PATCH_GRID = None    #TODO: implement
_C.MODEL.TRANS.PATCH_SIZE = 16
_C.MODEL.TRANS.HIDDEN_SIZE = 768  # 768(Base), 1024(Large), 1280(Huge)
_C.MODEL.TRANS.MLP_RATIO = 4
_C.MODEL.TRANS.NUM_HEADS = 12      # 12(Base), 16(Large), 16(Huge)
_C.MODEL.TRANS.NUM_LAYERS = 12     # 12(Base), 24(Large), 32(Huge)
_C.MODEL.TRANS.QKV_BIAS = True


# MLA Decoder setting
_C.MODEL.MLA = CN()
#_C.MODEL.MLA.MLA_INDEX = [2, 5, 8, 11]   # Base: [2, 5, 8, 11]; Large: [5, 11, 17, 23] 
_C.MODEL.MLA.MLA_CHANNELS = 256
_C.MODEL.MLA.MLAHEAD_CHANNELS=128
_C.MODEL.MLA.AUXIHEAD = False
_C.MODEL.MLA.MLAHEAD_ALIGN_CORNERS = False


# PUP and Naive Decoder setting
_C.MODEL.PUP = CN()
_C.MODEL.PUP.INPUT_CHANNEL = 1024
_C.MODEL.PUP.NUM_CONV = 4
_C.MODEL.PUP.NUM_UPSAMPLE_LAYER = 4
_C.MODEL.PUP.CONV3x3_CONV1x1 = True
_C.MODEL.PUP.ALIGN_CORNERS = False

# Auxi PUP and Naive Decoder setting
_C.MODEL.AUXPUP = CN()
_C.MODEL.AUXPUP.INPUT_CHANNEL = 1024
_C.MODEL.AUXPUP.NUM_CONV = 2
_C.MODEL.AUXPUP.NUM_UPSAMPLE_LAYER = 2
_C.MODEL.AUXPUP.CONV3x3_CONV1x1 = True
_C.MODEL.AUXPUP.ALIGN_CORNERS = False


# Auxilary Segmentation Head setting
_C.MODEL.AUX = CN()
_C.MODEL.AUX.AUXIHEAD = True
_C.MODEL.AUX.AUXHEAD_ALIGN_CORNERS = False

# training settings
_C.TRAIN = CN()
_C.TRAIN.LAST_EPOCH = 0
_C.TRAIN.BASE_LR = 0.001 #0.003 for pretrain # 0.03 for finetune
_C.TRAIN.END_LR = 1e-4
_C.TRAIN.DECODER_LR_COEF = 1.0
_C.TRAIN.GRAD_CLIP = 1.0
_C.TRAIN.ITERS = 80000
_C.TRAIN.WEIGHT_DECAY = 0.0 # 0.0 for finetune
_C.TRAIN.POWER=0.9
_C.TRAIN.DECAY_STEPS= 80000

_C.TRAIN.LR_SCHEDULER = CN()
_C.TRAIN.LR_SCHEDULER.NAME = 'PolynomialDecay'
_C.TRAIN.LR_SCHEDULER.MILESTONES = "30, 60, 90" # only used in StepLRScheduler
_C.TRAIN.LR_SCHEDULER.DECAY_EPOCHS = 30 # only used in StepLRScheduler
_C.TRAIN.LR_SCHEDULER.DECAY_RATE = 0.1 # only used in StepLRScheduler

_C.TRAIN.OPTIMIZER = CN()
_C.TRAIN.OPTIMIZER.NAME = 'SGD'
_C.TRAIN.OPTIMIZER.EPS = 1e-8
_C.TRAIN.OPTIMIZER.BETAS = (0.9, 0.999)  # for adamW
_C.TRAIN.OPTIMIZER.MOMENTUM = 0.9

# val settings
_C.VAL = CN()
_C.VAL.MULTI_SCALES_VAL = False
_C.VAL.IMAGE_BASE_SIZE = 520 # 520 for pascal context
_C.VAL.CROP_SIZE = [480,480]
_C.VAL.STRIDE_SIZE = [320,320]

# misc
_C.SAVE_DIR = "./output"
_C.KEEP_CHECKPOINT_MAX = 3
_C.TAG = "default"
_C.SAVE_FREQ_CHECKPOINT = 10 # freq to save chpt
_C.LOGGING_INFO_FREQ = 5 # freq to logging info
_C.VALIDATE_FREQ = 20 # freq to do validation
_C.SEED = 0
_C.EVAL = False # run evaluation only
_C.LOCAL_RANK = 0


def _update_config_from_file(config, cfg_file):
    config.defrost()
    with open(cfg_file, 'r') as infile:
        yaml_cfg = yaml.load(infile, Loader=yaml.FullLoader)
    for cfg in yaml_cfg.setdefault('BASE', ['']):
        if cfg:
            _update_config_from_file(
                config, os.path.join(os.path.dirname(cfg_file), cfg)
            )
    print('merging config from {}'.format(cfg_file))
    config.merge_from_file(cfg_file)
    config.freeze()


def update_config(config, args):
    if args.cfg:
        _update_config_from_file(config, args.cfg)
    config.defrost()

    config.freeze()
    return config


def get_config():
    config = _C.clone()
    return config

    


